openslides_backend/action/sort_generic.py

```diff
@@ -0,0 +1,143 @@
+from typing import Any, Dict, Iterable, List, Set
+
+import fastjsonschema
+
+from ..models.base import Model
+from ..shared.exceptions import ActionException
+from ..shared.filters import FilterOperator
+from ..shared.interfaces import Event, WriteRequestElement
+from ..shared.patterns import FullQualifiedId
+from ..shared.schema import schema_version
+from .base import BaseAction, DataSet
+
+sort_node_schema = {
+    "$schema": schema_version,
+    "title": "Sort node schema",
+    "description": "A node inside a sort tree.",
+    "type": "object",
+    "properties": {
+        "id": {
+            "description": "The id of the instance.",
+            "type": "integer",
+            "minimum": 1,
+        },
+        "children": {
+            "type": "array",
+            "items": {},  # TODO: Add recursive sort_node_schema here, then remove extra check in TreeSortMixin.
+            "minItems": 1,
+            "uniqueItems": True,
+        },
+    },
+    "required": ["id"],
+    "additionalProperties": False,
+}
+
+validate_sort_node = fastjsonschema.compile(sort_node_schema)
+
+
+class TreeSortMixin(BaseAction):
+    """
+    Provides an action mixin for sorting a model tree.
+    """
+
+    model: Model
+
+    def sort_tree(
+        self,
+        nodes: List,
+        meeting_id: int,
+        weight_key: str,
+        parent_id_key: str,
+        children_ids_key: str,
+    ) -> DataSet:
+        """
+        Sorts the all model objects represented in a tree of ids. The request
+        data should be a list (the root) of all main models. Each node is a dict
+        with an id and optional children. Every id has to be given.
+
+        This function traverses this tree in preorder to assign the weight.
+        """
+        # TODO: Check if instances exist in DB and is not deleted. Ensure that meta_deleted field is added to locked_fields.
+
+        # Get all item ids to verify, that the user send all ids.
+        filter = FilterOperator("meeting_id", "=", meeting_id)
+        db_instances = self.database.filter(
+            collection=self.model.collection,
+            filter=filter,
+            mapped_fields=["id"],
+            lock_result=True,
+        )
+        all_model_ids = set(map(int, db_instances.keys()))
```

Comment:
Please move this fix to the datastore adapter since it will effect all actions which use the filter method.

Reply:
Add a circle test and move the int fix to the datastore adapter.
Updated the tests.